### PR TITLE
Allows Pasvest, Marine Harness, and M10E helmet to be recolored.

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request

Allows Pas-11 Vest, Marine Harness, and M10E helmet to be recolored using Facepaint. This replaces the Map Based coloring for the armor pieces.

## Why It's Good For The Game

More cosmetic choices are always good.

## Changelog
:cl:
add: Allows Pas-11 Vest, Marine Harness, and M10E helmet to be recolored using facepaint like the Jaeger armor pieces.
/:cl:

### CODE REVIEW

**Comment by hyper2snyper** [comment]
why did it ask for a review?

**Comment by SkeletalElite** [comment]
> why did it ask for a review?

because bimmer is a code owner for icons, you touch icons, it automatically requests review from him

**Comment by hyper2snyper** [comment]
> > why did it ask for a review?
> 
> because bimmer is a code owner for icons, you touch icons, it automatically requests review from him

ah thanks, didnt know that.

**Comment by revelation-8468** [comment]
yes please

**Comment by hyper2snyper** [comment]
> yes please

definitely, now i dont have to wait for ice colony to do my cool medic outfit

**Comment by Sulaboy** [comment]
Looking at the code does this remove map variants from their original vending? Like will every map vend the same color armor for pas vests, or will they still bend as map specific?

**Review by Sulaboy — COMMENTED** [review]
`	-	/obj/item/weapon/twohanded, `
Line 34 marine_armor.do.

I think you snipped out it fitting two handed melees.

**Comment by hyper2snyper** [comment]
> Looking at the code does this remove map variants from their original vending? Like will every map vend the same color armor for pas vests, or will they still bend as map specific?

When they vend, they will be their standard variant. If someone wants to change its color they would just have to use the facepaint.

**Comment by hyper2snyper** [comment]
> ` - /obj/item/weapon/twohanded, `
> Line 34 marine_armor.do.
> 
> I think you snipped out it fitting two handed melees.

I dont remember doing any of those, in any case, they shouldnt be removed so ill fix it.

**Comment by Sulaboy** [comment]
Would it be possible to default it to black?

**Review by SplinterGP — CHANGES_REQUESTED** [review]
this needs to have copypasta reduced, it's all the same code for all the procs, possibly made an element

**Comment by hyper2snyper** [comment]
> Would it be possible to default it to black?

I don't see why that would be a problem. Black does look better.

**Comment by hyper2snyper** [comment]
> this needs to have copypasta reduced, it's all the same code for all the procs, possibly made an element

Do you mean to make it only a single proc to color all three?

**Comment by SplinterGP** [comment]
> 
> 
> > this needs to have copypasta reduced, it's all the same code for all the procs, possibly made an element
> 
> Do you mean to make it only a single proc to color all three?

Yep, as explained in discord elements can also be used

**Review by TheBimmer231 — APPROVED** [review]

**Comment by SplinterGP** [comment]
> 
> 
> So I created an element to handle color variants with the help of splinter (thanks much splinter)
> 
> Some testing found that the coloring code is being overwritten or something, so it doesn't work. (Not yet at least)
> 
> I'm considering making the element check for compatibility by looking at the colors list in each object and activating only if the list isn't null, that would allow me to initialize all the armors or whatnot with the element and just fill out the color list of the ones that have variants. (Was an idea I had, it might not work)
> 
> I'll do the autodoc and comments and such once the code actually works.
i'll talk about this later with ya

File: code/datums/elements/coloring.dm [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	var/colors = source.color_list
+	var/obj/item/facepaint/paint = attacked_by
+	if(paint.uses < 1)
+		to_chat(attacker, "<span class='warning'>\the [paint] is out of color!</span>")
+		return
+	
+	INVOKE_ASYNC(src, .proc/attackby, source, paint, attacker, colors, params)
+	return COMPONENT_NO_AFTERATTACK
+
+/datum/element/color/proc/attackby(obj/source, obj/item/attacked_by, mob/attacker, var/colors, params)
```
Comment (SplinterGP) [inline code/datums/elements/coloring.dm:33]:
```suggestion
/datum/element/color/proc/attackby(obj/source, obj/item/facepaint/paint, mob/attacker, var/colors, params) 
```

File: code/datums/elements/coloring.dm [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	if(paint.uses < 1)
+		to_chat(attacker, "<span class='warning'>\the [paint] is out of color!</span>")
+		return
+	
+	INVOKE_ASYNC(src, .proc/attackby, source, paint, attacker, colors, params)
+	return COMPONENT_NO_AFTERATTACK
+
+/datum/element/color/proc/attackby(obj/source, obj/item/attacked_by, mob/attacker, var/colors, params) 
+
+	var/obj/item/facepaint/paint = attacked_by
```
Comment (SplinterGP) [inline code/datums/elements/coloring.dm:35]:
```suggestion
```

File: code/game/objects/objs.dm [hunk 3]
```diff
@@ -32,6 +32,9 @@
 	///Optimization for dynamic explosion block values, for things whose explosion block is dependent on certain conditions.
 	var/real_explosion_block
 
+	///Variable for colors to be used if the object is given the capability to be recolored using facepaint.
```
Comment (SplinterGP) [inline code/game/objects/objs.dm:35]:
```suggestion
	///lazylist with colors to be used if the object is given the capability to be recolored using the recoloring element.
```

File: code/game/objects/objs.dm [hunk 4]
```diff
@@ -32,6 +32,9 @@
...
+	var/color_list = null
```
Comment (SplinterGP) [inline code/game/objects/objs.dm:36]:
```suggestion
	var/color_list
```

File: code/datums/elements/coloring.dm [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	var/new_color = tgui_input_list(attacker, "Pick a color", "Pick color", colors)
+	if(!new_color)
+		return
+	if(!do_after(attacker, 1 SECONDS, TRUE, src, BUSY_ICON_GENERIC))
+		return
+	
+	var/default_icon = initial(source.icon_state)
+	source.icon_state = "[default_icon]_[new_color]"
```
Comment (SplinterGP) [inline code/datums/elements/coloring.dm:43]:
```suggestion
	source.icon_state = "[initial(source.icon_state)]_[new_color]"
```

**Review by SplinterGP — CHANGES_REQUESTED** [review]

**Comment by Guidesu** [comment]
noooooooooooooooo!!

**Pull request closed without merging**